apps/prairielearn/python/question_phases.py

```diff
@@ -0,0 +1,187 @@
+    """A string consisting of `question.html` with Mustache templating applied."""
+
+    elements: Dict[str, ElementInfo]
+    """A dict mapping an element name to information about them."""
+
+    element_extensions: Dict[str, Dict[str, Dict]]
+    """A dict mapping an element name to a dict of extensions for that element."""
+
+    course_path: str
+    """The path to the course directory."""
+
+
+def filelike_to_string(filelike: Any) -> str:
+    # if val is None, replace it with empty string
+    if filelike is None:
+        filelike = ""
+
+    # if val is a file-like object, read whatever is inside
+    if isinstance(filelike, io.IOBase):
+        filelike.seek(0)
+        filelike = filelike.read()
+
+    # if val is a string, treat it as utf-8
+    if isinstance(filelike, str):
+        filelike = bytes(filelike, "utf-8")
+
+    # if this next call does not work, it will throw an error, because
+    # the thing returned by file() does not have the correct format
+    return base64.b64encode(filelike).decode()
+
+
+def process(
+    phase: Phase, data: dict, context: RenderContext
+) -> Tuple[Optional[str], Set[str]]:
+    html = context["html"]
+    elements = context["elements"]
+    element_extensions = context["element_extensions"]
+    course_path = context["course_path"]
+
+    # This will track which elements have been processed.
+    processed_elements: Set[str] = set()
+
+    # If we're in the `render` phase, we'll eventually capture the HTML here.
+    # If we're in the `file` phase, we'll capture file data here.
+    # Otherwise, this will remain `None`.
+    result = None
+
+    total_time = 0
+
+    def process_element(element: lxml.html.HtmlElement) -> Optional[str]:
+        nonlocal result
+
+        if element.tag not in elements:
+            return element
+
+        try:
+            start = time.time()
+            processed_elements.add(element.tag)
+
+            element_info = elements[element.tag]
+            element_type = element_info["type"]
+            element_name = element_info["name"]
+            if element_type == "core":
+                element_path = CORE_ELEMENTS_PATH / element_name
+            elif element_type == "course":
+                element_path = pathlib.Path(course_path) / "elements" / element_name
+            else:
+                assert_never(element_type)
+            element_controller = element_info["controller"]
+            element_controller_path = element_path / element_controller
+
+            # Set the element directory as the current working directory.
+            os.chdir(element_path)
+
+            # Update the path to include the appropriate directories.
+            sys.path = copy.copy(SAVED_PATH)
+            sys.path.insert(0, str(PYTHON_PATH))
+            if element_info["type"] == "course":
+                sys.path.insert(0, str(pathlib.Path(course_path) / "serverFilesCourse"))
+            sys.path.insert(0, str(element_path))
+
+            mod = {}
```

Comment:
Can this dict have a type annotation? Something like `dict[Phase, Callable[[str, dict], Optiona[str]]]`.

Reply:
This again feels like lying to the type-checker; we don't actually know *what* the module will be.